server/jetstream_api.go

```diff
@@ -2341,6 +2361,92 @@ func (s *Server) jsLeaderServerStreamMoveRequest(sub *subscription, c *client, _
 
 	cfg.Placement = origPlacement
 
+	s.Noticef("Requested move of: R=%d stream %s in account %s from old peer set %+v to new peer set %+v",
+		cfg.Replicas, streamName, accName, s.peerSetToNames(currPeers), s.peerSetToNames(peers))
+
+	// execute on a go routine for route or gateway, as this may block waiting from responses from other servers and
+	// therefore block all route/gateway traffic.
+	if c.kind != ROUTER && c.kind != GATEWAY {
+		go s.jsClusteredStreamUpdateRequest(&ciNew, targetAcc.(*Account), subject, reply, rmsg, &cfg, peers)
+	} else {
+		s.jsClusteredStreamUpdateRequest(&ciNew, targetAcc.(*Account), subject, reply, rmsg, &cfg, peers)
+	}
+}
+
+// Request to have the metaleader move a stream on a peer to another
+func (s *Server) jsLeaderServerStreamCancelMoveRequest(sub *subscription, c *client, _ *Account, subject, reply string, rmsg []byte) {
+	if c == nil || !s.JetStreamEnabled() {
+		return
+	}
+
+	ci, acc, _, msg, err := s.getRequestInfo(c, rmsg)
+	if err != nil {
+		s.Warnf(badAPIRequestT, msg)
+		return
+	}
+
+	js, cc := s.getJetStreamCluster()
+	if js == nil || cc == nil || cc.meta == nil {
+		return
+	}
+
+	// Extra checks here but only leader is listening.
+	js.mu.RLock()
+	isLeader := cc.isLeader()
+	js.mu.RUnlock()
+
+	if !isLeader {
+		return
+	}
+
+	var resp = JSApiStreamUpdateResponse{ApiResponse: ApiResponse{Type: JSApiStreamUpdateResponseType}}
+
+	accName := tokenAt(subject, 6)
+	streamName := tokenAt(subject, 7)
+
+	targetAcc, ok := s.accounts.Load(accName)
+	if !ok {
+		resp.Error = NewJSNoAccountError()
+		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	streamFound := false
+	cfg := StreamConfig{}
+	currPeers := []string{}
+	js.mu.Lock()
+	streams, ok := cc.streams[accName]
+	if ok {
+		sa, ok := streams[streamName]
+		if ok {
+			cfg = *sa.Config
+			streamFound = true
+			currPeers = sa.Group.Peers
+		}
+	}
+	js.mu.Unlock()
+
+	if !streamFound {
+		resp.Error = NewJSStreamNotFoundError()
+		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	if len(currPeers) <= cfg.Replicas {
+		resp.Error = NewJSStreamMoveNotInProgressError()
+		s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	// make sure client is scoped to requested account
+	ciNew := *(ci)
+	ciNew.Account = streamName
+
+	peers := currPeers[:cfg.Replicas]
+
+	s.Noticef("Requested cancel of move: R=%d stream %s in account %s from old peer set %+v to new peer set %+v",
+		cfg.Replicas, streamName, accName, s.peerSetToNames(currPeers), s.peerSetToNames(peers))
+
```

Comment:
Also, please use copyBytes(rmsg) below..

Reply:
fixed

Reply:
don't think copy is needed